### PR TITLE
feat: add user quiet hours schedule and restart requirement feature flag

### LINKED ISSUE #6917: Template max TTL time bounds
Max TTL may cause disruption to users if they start their workspace i.e. late at night and then they get forcibly ejected during the workday the next day.

E.g. user starts their workspace at 11pm and it has a max TTL of 12 hours, they would be forcibly kicked out at 11am during work hours.

Solution: alongside the max TTL, apply a work hours "bound" to the computed workspace max deadline to wrap it to the end of the day if it occurs during the day.

E.g. user starts their workspace at 11pm and it has a max TTL of 12 hours, but has a bound of "6pm to 7am", the max deadline would be pushed forward to 6pm as it occurs outside the max deadline window

Obviously, this only affects max deadline. If the user is AFK then their workspace will be stopped at the default TTL or their custom TTL.

------

One issue is that having the template admin denote bounds doesn't work well for international teams using the same template. Perhaps we could allow users to set their own work hours (with a max duration per day defined as a deployment var, defaulting to 12 hours).

### PR DESCRIPTION
**Pull request by deansheather**
Adds feature flag `template_restart_requirement` which changes `max_deadline` to be calculated using a new "restart requirement" setting on the template, rendering the `max_ttl` field unused.

Adds a `quiet_hours_schedule` field to the users table which stores a daily cron expression e.g. `CRON_TZ=Australia/Sydney 2 0 * * *` (every day at 2am Sydney time).

Adds `restart_requirement_days_of_week` and `restart_requirement_weeks` to templates table. The first value is a uint8 for all 7 days of the week (and an unused bit) governing which days of the week the policy should require restarts on (in the user's quiet hours schedule and timezone). The second value says how many weeks between required restarts. 0-1 means weekly, 2 means fortnightly etc.

The week calculations are done as an offset from an "epoch" date, so every workspace from every customer will have the same n-week calculations for consistency.

Admins must configure the default schedule that applies to users who haven't configured a custom schedule yet, and can also configure the window duration. Users can set a custom schedule in any timezone. The cron expression must be a daily expression with a single time each day.

TODO:
- [x] Configurable default schedule as flag
    - [x] Feature is disabled until admin picks a default schedule
- [x] Configurable window duration as flag
- [x] Tests for schedule package
- [x] Tests for new APIs
- [x] Tests for provisionerdserver applying the expected values to builds

Closes #6917

### CODE REVIEW

File: coderd/schedule/autostop.go
Comment (deansheather) [inline coderd/schedule/autostop.go:118]: This algorithm might be too naive??? More research/testing required to ensure this generates correct dates

Comment (deansheather) [reply coderd/schedule/autostop.go]: Redid the algorithm completely since this algorithm fails if DST has passed once already during the year. The new algorithm is good in a bunch of timezones for the next 100 years (and this is tested).

**Review by coadler — COMMENTED** [review]
will get to the rest tomorrow

File: coderd/schedule/autostop.go
Comment (Emyrk) [inline coderd/schedule/autostop.go:22]: I think I get this, but commenting some examples might help illustrate

**Review by Emyrk — COMMENTED** [review]
:heart: the docs

I need a longer stretch to grock some of this time stuff. The comments and code look well formatted, will get back to this :+1:

File: coderd/schedule/autostop.go
Comment (deansheather) [reply coderd/schedule/autostop.go]: I added examples on how these times help to each of them, let me know if I need to make them better

File: coderd/schedule/autostop.go
Comment (Emyrk) [inline coderd/schedule/autostop.go:326]: Do you have any example of why this would fail?

File: coderd/schedule/template.go
Comment (Emyrk) [inline coderd/schedule/template.go:52]: Why store as int64 if 16 is the max value?

File: coderd/schedule/template.go
Comment (Emyrk) [inline coderd/schedule/template.go:71]: Should we make `r.DaysOfWeek&(1<<uint(i)) != 0` some helper func on it so `DaysMap` and `Days` use the same boolean func to decide to add it to their set?

**Review by Emyrk — COMMENTED** [review]
I have not manually tested this, but the time stuff all looks good :+1: 

The bitmap feels unnecessary at first, since we always convert it into and out of a list. But it does give us validation that it uses each day at most once, all days are in order, and there is a maximum size to the list. I think LoC wise it's more to use the bitmap.

Example is where we use it here: https://github.com/coder/coder/blob/87b065b8552a76ffce10a0c4945f472b5e0278d7/coderd/schedule/autostop.go#L157-L157. We just expand it into a map.

----

I have not looked at the schedule package much at all, so there are some things I am not sure about.

This has Sunday as the first day. Is this ok?

https://github.com/coder/coder/blob/87b065b8552a76ffce10a0c4945f472b5e0278d7/coderd/schedule/cron.go#L190-L190

File: coderd/schedule/autostop.go
Comment (deansheather) [reply coderd/schedule/autostop.go:248]: Daylight savings would cause the old algorithm to be off by the daylight savings offset after a single DST event has passed but not the second DST event in the timezone during that year.

The new algorithm operates on dates only and doesn't touch hours/minutes, so it avoids this issue. There are heaps of tests for this algorithm in `autostop_test.go`, it's tested in a bunch of popular timezones for the next 5200 weeks (aka. 100 years).

File: coderd/schedule/template.go
Comment (deansheather) [reply coderd/schedule/template.go:52]: I use bigint (int64) in the database in case we ever want to remove this limit or make it configurable for each deployment. 64 bit is probably overkill though, you're right.

File: codersdk/deployment.go
Comment (coadler) [inline codersdk/deployment.go:1604]: I wonder if this would be better served as two cron schedules; one for the beginning of the quiet hour window and one for the end. The current implementation forces each quiet hour window be the same length, but I could see it being helpful to have longer windows on the weekends or specific days.

**Comment by coadler** [comment]
Personally I like the bitmap for storing in the DB, it makes the Postgres side pretty simple without the need for arrays.

**Comment by Emyrk** [comment]
> Personally I like the bitmap for storing in the DB, it makes the Postgres side pretty simple without the need for arrays. It felt overkill at first, but I think it's a suitable tradeoff.

It's probably not worth changing. I could go either way in it.

**Review by coadler — APPROVED** [review]

File: coderd/schedule/autostop.go
Comment (Emyrk) [reply coderd/schedule/autostop.go:326]: This is my only open question.

File: coderd/schedule/template.go
Comment (Emyrk) [reply coderd/schedule/template.go:52]: Doesn't matter, was just curious

File: codersdk/deployment.go
Comment (Emyrk) [reply codersdk/deployment.go:1604]: Seems like that could be an iterative addition if required. 

If we end up supporting 2 cron schedules instead of this format, we could always convert this to the second cron schedule.

**Review by Emyrk — APPROVED** [review]
:+1: Just my one open comment on how something might fail. It's fine to keep the defensive programming, just curious because I don't see how it could.

File: codersdk/deployment.go
Comment (deansheather) [reply codersdk/deployment.go:1604]: I'm going to remove the duration for now since it's unused. It was used at one point early on in this PR to apply jitter to the max_deadline, but we are instead going to use it (in a future PR) to allow users to postpone the max_deadline by a given amount if they happen to be working when the quiet hours ticks over.

File: coderd/schedule/autostop.go
Comment (deansheather) [reply coderd/schedule/autostop.go:326]: To be honest, I'm not sure how this could fail, but the `GetMondayOfWeek` algorithm just adds `7 * n` days to the epoch which seems kinda naive to me. This is tested though for the next 100 years in all of those timezones so I think it's stable, I'll add a comment saying this is probably not needed but here just in case.

File: coderd/schedule/template.go
Comment (deansheather) [reply coderd/schedule/template.go:71]: We're not using Days() so I removed it

**Comment by deansheather** [comment]
> This has Sunday as the first day. Is this ok?

Yeah we're not using it in this code, that's used for the autostart scheduling.

File: coderd/schedule/autostop.go
Comment (Emyrk) [reply coderd/schedule/autostop.go]: Perfect. Was just making sure I wasn't missing something.

**Comment by bpmct** [comment]
I just deployed this, this doesn't include the restart requirement in the `template settings` form, right? That will be a separate PR?

Also, thoughts on another name instead of "restart requirement" since the action will be a stop, but the user could intervene and bump or restart if they wish to keep working?

- Max session length
- Auto-stop schedule (different than inactivity timeout)

Perhaps "restart requirement" could work, but didn't feel natural to me. I tried documenting it and couldn't find a natural way to explain it or show it in the UI. Feel free to make the final call here @deansheather 
https://github.com/coder/coder/compare/main...bpmct/scheduling-docs

**Comment by deansheather** [comment]
> Also, thoughts on another name instead of "restart requirement" since the action will be a stop, but the user could intervene and bump or restart if they wish to keep working?

In the interest of not making this PR wait for too much longer I'm going to merge it with the types and field names as they are with `restart_requirement` and we can rename them later once we have a good name.

I dislike "max session length" as you do not set a duration with this new method. If you were setting a duration in hours like the old max_ttl then it would make sense.

Auto-stop schedule doesn't convey that it's a required stop and the workspace can't stay on after this point. Auto-stop requirement could work though.

**Pull request merged**
